api/classes/publisher.js

```diff
@@ -79,72 +115,236 @@ function deleteFile(file, root) {
-    .then(function() {
-      log.info(
-        'Successfully published ' +
-        project.get('title') +
-        ' to ' + project.get('publish_url')
-      );
-    })
-    .catch(function(e) {
-      log.error(e);
+    .then(function(f) {
+      return Promise.map(f, function(toBePublished) {
+        modifyPublishedFiles(uploadFile, project, user, toBePublished);
+      });
+    });
+  })
+  .then(function() {
+    // Confirm we have files to be deleted
+    if (toBeDeleted.length === 0) {
+      return;
+    }
+
+    return Promise.map(toBeDeleted, function(model) {
+      return modifyPublishedFiles(deleteFileRemotely, project, user, model)
+      .then(function() {
+        return model.destroy();
+      });
+    });
+  })
+  .then(function() {
+    // Confirm we have files to be updated
+    if (toBeUpdated.length === 0) {
+      return;
+    }
+
+    return Promise.map(toBeUpdated, function(model) {
+      return modifyPublishedFiles(deleteFileRemotely, project, user, model)
+      .then(function() {
+        var originalFile = projectFiles[model.get('file_id')];
 
-      return Promise.reject(e);
+        return model.set({
+          path: originalFile.get('path'),
+          buffer: originalFile.get('buffer')
+        }).save()
+        .then(function() {
+          return modifyPublishedFiles(uploadFile, project, user, model);
+        });
+      });
     });
+  })
+  .then(function() {
+    return project.set({
+      publish_url: buildUrl(project, user),
+      published_id: publishedProject.get('id')
+    }).save();
+  })
+  .then(function() {
+    log.info(
+      'Successfully published ' +
+      project.get('title') +
+      ' to ' + project.get('publish_url')
+    );
+  })
+  .catch(reject);
 };
 
-exports.unpublish = function(project) {
-  return getUser(project)
-    .then(function(user) {
-      return modifyFiles(deleteFile, project, user);
-    })
-    .then(function() {
-      return project.set({ publish_url: null }).save();
-    })
-    .then(function() {
-      log.info(
-        'Successfully unpublished ' +
-        project.get('title')
-      );
-    })
-    .catch(function(e) {
-      log.error(e);
+exports.unpublish = function unpublish(project) {
+  return Promise.join(getUserForProject(project),
+                      getPublishedProject(project),
+                      function(user, publishedProject) {
+    // First, collect published files
+    return getPublishedFiles(publishedProject)
+    .then(function(files) {
+      // Then, remove them from s3
+      return modifyPublishedFiles(deleteFileRemotely, project, user, files)
+      .then(function() {
+        // Then, remove the reference from the parent project
+        return project.set({
+          published_id: null
+        }).save();
+      })
+      .then(function() {
+        // Finally, delete them from the DB
+        var deletePublishedProject = publishedProject.destroy();
+        var deletePublishedFiles = files.mapThen(function(file) {
```

Comment:
Do you need to do this? if the publishedFiles has a FK to publishedProject with cascade on delete enabled, it should automatically delete them shouldn't it?